jib-gradle-plugin/src/main/java/com/google/cloud/tools/jib/gradle/BuildImageTask.java

```diff
@@ -81,20 +86,45 @@ public void buildImage() throws InvalidImageReferenceException {
               .forToNotConfigured(
                   "'jib.to.image'", "build.gradle", "gradle jib --image <your image name>"));
     }
-    RegistryCredentials knownTargetRegistryCredentials = null;
-    Authorization toAuthorization =
-        ConfigurationPropertyValidator.getImageAuth(
+
+    ImageReference targetImage = ImageReference.parse(jibExtension.getTargetImage());
+
+    CredentialRetrieverFactory credentialRetrieverFactory =
+        CredentialRetrieverFactory.forImage(targetImage, gradleJibLogger);
+    Credential toCredential =
+        ConfigurationPropertyValidator.getImageCredential(
             gradleJibLogger,
             "jib.to.auth.username",
             "jib.to.auth.password",
             jibExtension.getTo().getAuth());
-    if (toAuthorization != null) {
-      knownTargetRegistryCredentials = new RegistryCredentials("jib.to.auth", toAuthorization);
+    RegistryCredentials knownTargetRegistryCredentials = null;
+    CredentialRetriever knownCredentialRetriever = null;
+    if (toCredential != null) {
+      knownTargetRegistryCredentials =
+          new RegistryCredentials(
+              "jib.to.auth",
+              Authorizations.withBasicCredentials(
+                  toCredential.getUsername(), toCredential.getPassword()));
+      knownCredentialRetriever = credentialRetrieverFactory.known(toCredential, "jib.to.auth");
     }
+    // Makes credential retriever list.
+    List<CredentialRetriever> credentialRetrievers = new ArrayList<>();
+    String credentialHelperSuffix = jibExtension.getFrom().getCredHelper();
+    if (credentialHelperSuffix != null) {
+      credentialRetrievers.add(
+          credentialRetrieverFactory.dockerCredentialHelper(credentialHelperSuffix));
+    }
+    if (knownCredentialRetriever != null) {
+      credentialRetrievers.add(knownCredentialRetriever);
+    }
+    credentialRetrievers.add(credentialRetrieverFactory.inferCredentialHelper());
+    credentialRetrievers.add(credentialRetrieverFactory.dockerConfig());
+
     ImageConfiguration targetImageConfiguration =
-        ImageConfiguration.builder(ImageReference.parse(jibExtension.getTargetImage()))
+        ImageConfiguration.builder(targetImage)
             .setCredentialHelper(jibExtension.getTo().getCredHelper())
             .setKnownRegistryCredentials(knownTargetRegistryCredentials)
+            .setCredentialRetrievers(credentialRetrievers)
```

Comment:
Is there any reason anymore to set all 3 of these, or do we just need `setCredentialRetrievers()`? Not necessarily in this PR, just in general.

Reply:
Yep, it will eventually be just `setCredentialRetrievers`. I'm just doing these changes piece by piece so we have the behavior does not change at each step.